planner/core/task.go

```diff
@@ -318,67 +521,86 @@ func (p *NominalSort) attach2Task(tasks ...task) task {
-		// If all columns in topN are from index plan, we can push it to index plan. Or we finish the index plan and
+		// If all columns in topN are from index plan, we push it to index plan, otherwise we finish the index plan and
 		// push it to table plan.
+		var pushedDownTopN *PhysicalTopN
 		if !copTask.indexPlanFinished && p.allColsFromSchema(copTask.indexPlan.Schema()) {
-			pushedDownTopN.SetChildren(copTask.indexPlan)
+			pushedDownTopN = p.getPushedDownTopN(copTask.indexPlan)
 			copTask.indexPlan = pushedDownTopN
 		} else {
-			// FIXME: When we pushed down a top-N plan to table plan branch in case of double reading. The cost should
-			// be more expensive in case of single reading, because we may execute table scan multi times.
 			copTask.finishIndexPlan()
-			pushedDownTopN.SetChildren(copTask.tablePlan)
+			pushedDownTopN = p.getPushedDownTopN(copTask.tablePlan)
 			copTask.tablePlan = pushedDownTopN
 		}
-		copTask.addCost(pushedDownTopN.getCost(t.count()))
+		copTask.addCost(pushedDownTopN.GetCost(inputCount, false))
 	}
-	t = finishCopTask(p.ctx, t)
-	t = attachPlan2Task(p, t)
-	t.addCost(p.getCost(t.count()))
-	return t
+	rootTask := finishCopTask(p.ctx, t)
+	rootTask.addCost(p.GetCost(rootTask.count(), true))
+	rootTask = attachPlan2Task(p, rootTask)
+	return rootTask
+}
+
+// GetCost computes the cost of projection operator itself.
+func (p *PhysicalProjection) GetCost(count float64) float64 {
+	cpuCost := count * cpuFactor
+	concurrency := float64(p.ctx.GetSessionVars().ProjectionConcurrency)
+	if concurrency <= 0 {
+		return cpuCost
+	}
+	cpuCost /= concurrency
+	concurrencyCost := (1 + concurrency) * concurrencyFactor
+	return cpuCost + concurrencyCost
 }
 
 func (p *PhysicalProjection) attach2Task(tasks ...task) task {
 	t := tasks[0].copy()
-	switch tp := t.(type) {
-	case *copTask:
-		// TODO: Support projection push down.
-		t = finishCopTask(p.ctx, t)
-		t = attachPlan2Task(p, t)
-		return t
-	case *rootTask:
-		return attachPlan2Task(p, tp)
+	if copTask, ok := t.(*copTask); ok {
+		// TODO: support projection push down.
+		t = finishCopTask(p.ctx, copTask)
 	}
-	return nil
+	t = attachPlan2Task(p, t)
+	t.addCost(p.GetCost(t.count()))
+	return t
 }
 
 func (p *PhysicalUnionAll) attach2Task(tasks ...task) task {
-	newTask := &rootTask{p: p}
-	newChildren := make([]PhysicalPlan, 0, len(p.children))
+	t := &rootTask{p: p}
+	childPlans := make([]PhysicalPlan, 0, len(tasks))
+	var childMaxCost float64
 	for _, task := range tasks {
 		task = finishCopTask(p.ctx, task)
-		newTask.cst += task.cost()
-		newChildren = append(newChildren, task.plan())
+		childCost := task.cost()
+		if childCost > childMaxCost {
+			childMaxCost = childCost
+		}
+		childPlans = append(childPlans, task.plan())
 	}
-	p.SetChildren(newChildren...)
-	return newTask
+	p.SetChildren(childPlans...)
+	// Children of UnionExec are executed in parallel.
+	t.cst = childMaxCost + float64((1+len(tasks)))*concurrencyFactor
```

Comment:
In `UnionAll`, if `len(children)` is very big, can we still use `childMaxCost` as its cost?

Reply:
If `len(children)` is very big, the execution time may be longer than the corresponding time of estimated `childMaxCost`, since CPU load would be higher then, while this actually should be reflected in "concurrency cost", now we are using a constant factor for `concurrencyFactor` because it is simple, we can optimize this to make "concurrency cost" be an exponential of concurrency number if we find it is necessary in the future.

Reply:
All right.